helix-core/src/main/java/org/apache/helix/task/TaskUtil.java

```diff
@@ -990,46 +1009,65 @@ public static void purgeExpiredJobs(String workflow, WorkflowConfig workflowConf
+          for (String job : expiredJobs) {
+            if (!TaskUtil.removeJob(manager.getHelixDataAccessor(), manager.getHelixPropertyStore(),
+                job)) {
+              failedJobRemovals.add(job);
+              LOG.warn("Failed to clean up expired and completed jobs from workflow " + workflow);
+            }
+            rebalanceScheduler.removeScheduledRebalance(job);
           }
-          rebalanceScheduler.removeScheduledRebalance(job);
-        }
 
-        // If the job removal failed, make sure we do NOT prematurely delete it from DAG so that the
-        // removal will be tried again at next purge
-        expiredJobs.removeAll(failedJobRemovals);
+          // If the job removal failed, make sure we do NOT prematurely delete it from DAG so that
+          // the removal will be tried again at next purge
+          expiredJobs.removeAll(failedJobRemovals);
 
-        if (!TaskUtil.removeJobsFromDag(manager.getHelixDataAccessor(), workflow, expiredJobs,
-            true)) {
-          LOG.warn("Error occurred while trying to remove jobs + " + expiredJobs
-              + " from the workflow " + workflow);
+          if (!TaskUtil.removeJobsFromDag(manager.getHelixDataAccessor(), workflow, expiredJobs,
+              true)) {
+            LOG.warn("Error occurred while trying to remove jobs + " + expiredJobs
+                + " from the workflow " + workflow);
+          }
         }
-
-        if (expiredJobs.size() > 0) {
-          // Update workflow context will be in main pipeline not here. Otherwise, it will cause
-          // concurrent write issue. It is possible that jobs got purged but there is no event to
-          // trigger the pipeline to clean context.
-          HelixDataAccessor accessor = manager.getHelixDataAccessor();
-          List<String> resourceConfigs =
-              accessor.getChildNames(accessor.keyBuilder().resourceConfigs());
-          if (resourceConfigs.size() > 0) {
-            RebalanceUtil.scheduleOnDemandPipeline(manager.getClusterName(), 0L);
-          } else {
-            LOG.warn(
-                "No resource config to trigger rebalance for clean up contexts for" + expiredJobs);
+        if (!misconfiguredJobs.isEmpty()) {
+          LOG.info("Purge jobs " + misconfiguredJobs + " from queue " + workflow);
+          for (String job : misconfiguredJobs) {
+            if (!TaskUtil.removeJob(manager.getHelixDataAccessor(), manager.getHelixPropertyStore(),
+                job)) {
+              LOG.warn("Failed to clean up misconfigured jobs from workflow " + workflow);
+            }
+            rebalanceScheduler.removeScheduledRebalance(job);
           }
+          if (!TaskUtil.removeJobsFromDag(manager.getHelixDataAccessor(), workflow,
+              misconfiguredJobs, true)) {
+            LOG.warn("Error occurred while trying to remove jobs + " + misconfiguredJobs
+                + " from the workflow " + workflow);
+          }
+        }
+      }
```

Comment:
This is exactly same code as remove expired jobs. Two options here: 1) add the misconfig jobs to expired job and 2) move the logic to a generic private function to reduce the redundant code.

Reply:
Your point is kind of true. However, we treat them differently. In one expired jobs we try to remove them. For jobs without config, we remove them from the dag. There is no point in trying to remove jobs without config because they do not have any config and remove will fail.